triton-isa/src/op_stack.rs

```diff
@@ -183,13 +182,69 @@ impl Index<usize> for OpStack {
     }
 }
 
+fn compute_range(start: usize, end: usize, len: usize) -> (usize, usize) {
+    match (start, end) {
+        (0, end) if end == len => (0, len),
+        (0, end) => (len - end, len),
+        _ => (len - end, len - start),
+    }
+}
+
+impl Index<Range<usize>> for OpStack {
+    type Output = [BFieldElement];
+
+    fn index(&self, range: Range<usize>) -> &Self::Output {
+        let (start, end) = compute_range(range.start, range.end, self.stack.len());
+        let mut stack = self.stack[start..end].to_vec();
+        stack.reverse();
+        Box::leak(Box::new(stack))
```

Comment:
That's a creative way of solving this. :smiley: However, I feel uncomfortable with consciously introducing memory leaks.

The approach you take here indicates to me that implementing `Index<Range<_>> for OpStack` requires refactoring `OpStack` such that the internal field `stack` is always in the correct order, for example using a `VecDeque`. It's probably a good idea to make the field non-public at the same time. Please let me know if this is too much scope creep. :relieved:

Reply:
Introduction of` VecDeque` sounds a good idea.

```
pub struct OpStack {
    pub stack: VecDeque<BFieldElement>,

    underflow_io_sequence: Vec<UnderflowIO>,
}

impl OpStack {
    pub fn new(program_digest: Digest) -> Self {
        let mut stack = VecDeque::with_capacity(OpStackElement::COUNT);
        stack.extend(program_digest.values());

        Self {
            stack,
            underflow_io_sequence: vec![],
        }
    }
```
In this case we do not reverse `program_digest` and set `self.stack[0]` is a top of the stack.

Additionally, we can update `stack.push, stack.pop `to add new item to the beginning of the stack and remove item from the beginning of the stack, respectively. `Vec<BFieldElement>` lacks this functionality.

```
  pub fn push(&mut self, element: BFieldElement) {
        self.stack.push_front(element);
        self.record_underflow_io(UnderflowIO::Write);
    }

    pub fn pop(&mut self) -> Result<BFieldElement> {
        self.record_underflow_io(UnderflowIO::Read);
        self.stack.pop_front().ok_or(OpStackError::TooShallow)
    }

    pub fn insert(&mut self, index: OpStackElement, element: BFieldElement) {
        let insertion_index = usize::from(index);
        self.stack.insert(insertion_index, element);
        self.record_underflow_io(UnderflowIO::Write);
    }
```
`Index<Range<usize>> for OpStack` can be implemented like that :
```
impl Index<Range<usize>> for OpStack {
    type Output = [BFieldElement];

    fn index(&self, range: Range<usize>) -> &Self::Output {
        &self.stack.as_slices().0[range.start..range.end]
    }
}
```

I wonder if this make sense.